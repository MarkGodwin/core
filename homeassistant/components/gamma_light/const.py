"""Constants for the Switch as X integration."""

from typing import Final

DOMAIN: Final = "gamma_light"
MIN_BRIGHTNESS: Final = "min_brightness"
GAMMA: Final = "gamma"
