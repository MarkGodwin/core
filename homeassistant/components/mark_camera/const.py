"""Constants for the generic (IP Camera) integration."""

DOMAIN = "mark_camera"
DEFAULT_NAME = "Mark's Generic Camera"
CONF_CONFIRMED_OK = "confirmed_ok"
CONF_CONTENT_TYPE = "content_type"
CONF_LIMIT_REFETCH_TO_URL_CHANGE = "limit_refetch_to_url_change"
CONF_STILL_IMAGE_URL = "still_image_url"
CONF_STREAM_SOURCE = "stream_source"
CONF_FRAMERATE = "framerate"
CONF_MOTION_SENSOR = "motion_sensor_id"
CONF_DOORBELL_SENSOR = "doorbell_sensor_id"
GET_IMAGE_TIMEOUT = 10

DEFAULT_USERNAME = None
DEFAULT_PASSWORD = None
DEFAULT_IMAGE_URL = None
DEFAULT_STREAM_SOURCE = None
