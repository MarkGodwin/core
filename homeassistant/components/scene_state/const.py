"""Constants for the Scene State integration."""

DOMAIN = "scene_state"
