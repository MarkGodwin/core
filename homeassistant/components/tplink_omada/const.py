"""Constants for the TP-Link Omada integration."""

DOMAIN = "tplink_omada"

CONF_SITE = "site"

OPT_DEVICE_TRACKER = "device_tracker"
OPT_TRACKED_CLIENTS = "tracked_clients"
OPT_SCANNED_CLIENTS = "scanned_clients"
OPT_TRACKER_POLL_INTERVAL = "tracker_poll_interval"

DEFAULT_TRACKER_POLL_INTERVAL = 300
