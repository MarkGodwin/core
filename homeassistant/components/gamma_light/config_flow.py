"""Config flow for gamma_light."""

from homeassistant.core import HomeAssistant
from homeassistant.const import Platform
from homeassistant.helpers import config_entry_flow
from homeassistant.helpers import entity_registry

from .const import DOMAIN


async def _async_has_devices(hass: HomeAssistant) -> bool:
    """Return if there are devices that can be discovered."""
    # TODO Check if there are any devices that can be discovered in the network.

    registry = entity_registry.async_get(hass)

    potential_lights = [
        entry
        for entry in registry.entities.values()
        if entry.domain == Platform.LIGHT and entry.platform != DOMAIN
    ]

    # Exclude any lights we have already wrapped

    return len(potential_lights) > 0
    # devices = await hass.async_add_executor_job(my_pypi_dependency.discover)
    # return len(devices) > 0


config_entry_flow.register_discovery_flow(DOMAIN, "gamma_light", _async_has_devices)
