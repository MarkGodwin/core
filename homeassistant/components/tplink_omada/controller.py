"""Controller for sharing Omada API coordinators between platforms."""

import dataclasses
from dataclasses import dataclass
from typing import Any

from tplink_omada_client import OmadaSiteClient
from tplink_omada_client.devices import OmadaSwitch

from homeassistant.core import HomeAssistant

from .const import DEFAULT_TRACKER_POLL_INTERVAL
from .coordinator import (
    OmadaClientsCoordinator,
    OmadaGatewayCoordinator,
    OmadaSwitchPortCoordinator,
)


@dataclass
class OmadaIntegrationOptions:
    """Options for the Omada integration."""

    device_tracker: bool = False
    tracked_clients: list[str] = []
    scanned_clients: list[str] = []
    tracker_poll_interval: int = DEFAULT_TRACKER_POLL_INTERVAL


class OmadaSiteController:
    """Controller for the Omada SDN site."""

    _gateway_coordinator: OmadaGatewayCoordinator | None = None
    _initialized_gateway_coordinator = False
    _clients_coordinator: OmadaClientsCoordinator | None = None

    def __init__(
        self,
        hass: HomeAssistant,
        omada_client: OmadaSiteClient,
        options: dict[str, Any],
    ) -> None:
        """Create the controller."""
        self._hass = hass
        self._omada_client = omada_client
        self._options = self._get_omada_options(options)

        self._switch_port_coordinators: dict[str, OmadaSwitchPortCoordinator] = {}

    def _get_omada_options(self, options: dict[str, Any]) -> OmadaIntegrationOptions:
        return OmadaIntegrationOptions(
            **{
                k: v
                for k, v in options.items()
                if k in dataclasses.fields(OmadaIntegrationOptions)
            }
        )

    @property
    def options(self) -> OmadaIntegrationOptions:
        """Get the options for the integration."""
        return self._options

    @property
    def omada_client(self) -> OmadaSiteClient:
        """Get the connected client API for the site to manage."""
        return self._omada_client

    def get_switch_port_coordinator(
        self, switch: OmadaSwitch
    ) -> OmadaSwitchPortCoordinator:
        """Get coordinator for network port information of a given switch."""
        if switch.mac not in self._switch_port_coordinators:
            self._switch_port_coordinators[switch.mac] = OmadaSwitchPortCoordinator(
                self._hass, self._omada_client, switch
            )

        return self._switch_port_coordinators[switch.mac]

    async def get_gateway_coordinator(self) -> OmadaGatewayCoordinator | None:
        """Get coordinator for site's gateway, or None if there is no gateway."""
        if not self._initialized_gateway_coordinator:
            self._initialized_gateway_coordinator = True
            devices = await self._omada_client.get_devices()
            gateway = next((d for d in devices if d.type == "gateway"), None)
            if not gateway:
                return None

            self._gateway_coordinator = OmadaGatewayCoordinator(
                self._hass, self._omada_client, gateway.mac
            )

        return self._gateway_coordinator

    def get_clients_coordinator(self) -> OmadaClientsCoordinator:
        """Get coordinator for site's clients."""
        if not self._clients_coordinator:
            self._clients_coordinator = OmadaClientsCoordinator(
                self._hass,
                self._omada_client,
                self._options,
            )

        return self._clients_coordinator
