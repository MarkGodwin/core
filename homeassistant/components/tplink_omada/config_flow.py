"""Config flow for TP-Link Omada integration."""

from __future__ import annotations

from collections.abc import Mapping
import logging
import re
from types import MappingProxyType
from typing import Any, NamedTuple
from urllib.parse import urlsplit

from aiohttp import CookieJar
from tplink_omada_client import OmadaClient, OmadaSite
from tplink_omada_client.clients import OmadaWirelessClient
from tplink_omada_client.exceptions import (
    ConnectionFailed,
    LoginFailed,
    OmadaClientException,
    UnsupportedControllerVersion,
)
import voluptuous as vol

from homeassistant.config_entries import (
    ConfigEntry,
    ConfigFlow,
    ConfigFlowResult,
    OptionsFlow,
    OptionsFlowWithConfigEntry,
)
from homeassistant.const import (
    CONF_HOST,
    CONF_PASSWORD,
    CONF_USERNAME,
    CONF_VERIFY_SSL,
    UnitOfTime,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import selector
from homeassistant.helpers.aiohttp_client import (
    async_create_clientsession,
    async_get_clientsession,
)

from .const import (
    CONF_SITE,
    DEFAULT_TRACKER_POLL_INTERVAL,
    DOMAIN,
    OPT_DEVICE_TRACKER,
    OPT_SCANNED_CLIENTS,
    OPT_TRACKED_CLIENTS,
    OPT_TRACKER_POLL_INTERVAL,
)
from .controller import OmadaSiteController

_LOGGER = logging.getLogger(__name__)

_STEP_DEVICE_TRACKER = "device_tracker"

STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_HOST): str,
        vol.Required(CONF_VERIFY_SSL, default=True): bool,
        vol.Required(CONF_USERNAME): str,
        vol.Required(CONF_PASSWORD): str,
    }
)


async def create_omada_client(
    hass: HomeAssistant, data: MappingProxyType[str, Any]
) -> OmadaClient:
    """Create a TP-Link Omada client API for the given config entry."""

    host: str = data[CONF_HOST]
    verify_ssl = bool(data[CONF_VERIFY_SSL])

    if not host.lower().startswith(("http://", "https://")):
        host = "https://" + host
    host_parts = urlsplit(host)
    if (
        host_parts.hostname
        and re.fullmatch(r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}", host_parts.hostname)
        is not None
    ):
        # TP-Link API uses cookies for login session, so an unsafe cookie jar is required for IP addresses
        websession = async_create_clientsession(
            hass, cookie_jar=CookieJar(unsafe=True), verify_ssl=verify_ssl
        )
    else:
        websession = async_get_clientsession(hass, verify_ssl=verify_ssl)

    username = data[CONF_USERNAME]
    password = data[CONF_PASSWORD]

    return OmadaClient(host, username, password, websession=websession)


class HubInfo(NamedTuple):
    """Discovered controller information."""

    controller_id: str
    name: str
    sites: list[OmadaSite]


async def _validate_input(hass: HomeAssistant, data: dict[str, Any]) -> HubInfo:
    """Validate the user input allows us to connect."""

    client = await create_omada_client(hass, MappingProxyType(data))
    controller_id = await client.login()
    name = await client.get_controller_name()
    sites = await client.get_sites()

    return HubInfo(controller_id, name, sites)


class TpLinkOmadaConfigFlow(ConfigFlow, domain=DOMAIN):
    """Handle a config flow for TP-Link Omada."""

    VERSION = 1

    def __init__(self) -> None:
        """Create the config flow for a new integration."""
        self._omada_opts: dict[str, Any] = {}
        self._sites: list[OmadaSite] = []
        self._controller_name = ""

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""

        errors: dict[str, str] = {}
        info = None
        if user_input is not None:
            info = await self._test_login(user_input, errors)

        if info is None or user_input is None:
            return self.async_show_form(
                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
            )

        await self.async_set_unique_id(info.controller_id)
        self._abort_if_unique_id_configured()

        self._omada_opts.update(user_input)
        self._sites = info.sites
        self._controller_name = info.name
        if len(self._sites) > 1:
            return await self.async_step_site()
        return await self.async_step_site({CONF_SITE: self._sites[0].id})

    async def async_step_site(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle step to select site to manage."""

        if user_input is None:
            schema = vol.Schema(
                {
                    vol.Required(CONF_SITE, "site"): selector.SelectSelector(
                        selector.SelectSelectorConfig(
                            options=[
                                selector.SelectOptionDict(value=s.id, label=s.name)
                                for s in self._sites
                            ],
                            multiple=False,
                            mode=selector.SelectSelectorMode.DROPDOWN,
                        )
                    )
                }
            )

            return self.async_show_form(step_id="site", data_schema=schema)

        self._omada_opts.update(user_input)
        site_name = next(
            site for site in self._sites if site.id == user_input["site"]
        ).name
        display_name = f"{self._controller_name} ({site_name})"

        return self.async_create_entry(title=display_name, data=self._omada_opts)

    async def async_step_reauth(
        self, entry_data: Mapping[str, Any]
    ) -> ConfigFlowResult:
        """Perform reauth upon an API authentication error."""
        self._omada_opts = dict(entry_data)
        return await self.async_step_reauth_confirm()

    async def async_step_reauth_confirm(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Dialog that informs the user that reauth is required."""

        errors: dict[str, str] = {}

        if user_input is not None:
            self._omada_opts.update(user_input)
            info = await self._test_login(self._omada_opts, errors)

            if info is not None:
                # Auth successful - update the config entry with the new credentials
                entry = self.hass.config_entries.async_get_entry(
                    self.context["entry_id"]
                )
                assert entry is not None
                self.hass.config_entries.async_update_entry(
                    entry, data=self._omada_opts
                )
                await self.hass.config_entries.async_reload(entry.entry_id)
                return self.async_abort(reason="reauth_successful")

        return self.async_show_form(
            step_id="reauth_confirm",
            data_schema=vol.Schema(
                {
                    vol.Required(CONF_USERNAME): str,
                    vol.Required(CONF_PASSWORD): str,
                }
            ),
            errors=errors,
        )

    async def _test_login(
        self, data: dict[str, Any], errors: dict[str, str]
    ) -> HubInfo | None:
        try:
            info = await _validate_input(self.hass, data)
            if len(info.sites) > 0:
                return info
            errors["base"] = "no_sites_found"

        except ConnectionFailed:
            errors["base"] = "cannot_connect"
        except LoginFailed:
            errors["base"] = "invalid_auth"
        except UnsupportedControllerVersion:
            errors["base"] = "unsupported_controller"
        except OmadaClientException as ex:
            _LOGGER.error("Unexpected API error: %s", ex)
            errors["base"] = "unknown"
        except Exception:  # pylint: disable=broad-except
            _LOGGER.exception("Unexpected exception")
            errors["base"] = "unknown"
        return None

    @staticmethod
    @callback
    def async_get_options_flow(
        config_entry: ConfigEntry,
    ) -> OptionsFlow:
        """Create the options flow."""
        return OptionsFlowHandler(config_entry)


class OptionsFlowHandler(OptionsFlowWithConfigEntry):
    """Options flow for TP-Link Omada."""

    async def async_step_init(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Manage the top-level options."""

        self.options.update(user_input or {})
        device_tracker_enabled = self.options.get(OPT_DEVICE_TRACKER, False)

        if user_input is None:
            return self.async_show_form(
                step_id="init",
                data_schema=vol.Schema(
                    {
                        vol.Required(
                            OPT_DEVICE_TRACKER,
                            default=device_tracker_enabled,
                        ): bool
                    }
                ),
            )

        if device_tracker_enabled:
            return await self.async_step_device_tracker()

        return self.async_create_entry(data=self.options)

    async def async_step_device_tracker(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Manage the device tracker options."""
        errors = {}

        if user_input is not None:
            self.options.update(user_input)
            if (
                len(self.options[OPT_TRACKED_CLIENTS]) == 0
                and len(self.options[OPT_SCANNED_CLIENTS]) == 0
            ):
                errors[OPT_TRACKED_CLIENTS] = "no_clients_selected"

        if user_input is None or errors:
            controller: OmadaSiteController = self.hass.data[DOMAIN][
                self.config_entry.entry_id
            ]

            site_client = controller.omada_client
            # Get the known Wi-Fi clients, in alphabetical order moving un-named clients to the end of the list
            clients: list[OmadaWirelessClient] = sorted(
                [
                    client
                    async for client in site_client.get_known_clients()
                    if isinstance(client, OmadaWirelessClient)
                ],
                key=lambda c: (0 if c.name != c.mac else 1, c.name),
            )

            schema = vol.Schema(
                {
                    vol.Optional(
                        OPT_SCANNED_CLIENTS,
                        default=self.options.get(OPT_SCANNED_CLIENTS, []),
                    ): _get_clients_selector(clients),
                    vol.Optional(
                        OPT_TRACKED_CLIENTS,
                        default=self.options.get(OPT_TRACKED_CLIENTS, []),
                    ): _get_clients_selector(clients),
                    vol.Required(
                        OPT_TRACKER_POLL_INTERVAL,
                        default=self.options.get(
                            OPT_TRACKER_POLL_INTERVAL, DEFAULT_TRACKER_POLL_INTERVAL
                        ),
                    ): selector.NumberSelector(
                        selector.NumberSelectorConfig(
                            min=30,
                            max=600,
                            step=5,
                            unit_of_measurement=UnitOfTime.SECONDS,
                            mode=selector.NumberSelectorMode.SLIDER,
                        )
                    ),
                }
            )

            return self.async_show_form(
                step_id=_STEP_DEVICE_TRACKER, data_schema=schema, errors=errors
            )

        return self.async_create_entry(data=self.options)


def _get_clients_selector(
    clients: list[OmadaWirelessClient],
) -> selector.SelectSelector:
    return selector.SelectSelector(
        selector.SelectSelectorConfig(
            options=[
                selector.SelectOptionDict(value=c.mac, label=c.name) for c in clients
            ],
            multiple=True,
            mode=selector.SelectSelectorMode.DROPDOWN,
        )
    )
