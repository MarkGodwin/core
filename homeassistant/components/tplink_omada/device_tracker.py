"""Connected Wi-Fi device scanners for TP-Link Omada access points."""

from collections.abc import Callable
from dataclasses import dataclass
import logging

from tplink_omada_client.clients import OmadaWirelessClient
from tplink_omada_client.exceptions import RequestFailed

from homeassistant.components.device_tracker import (
    ScannerEntity,
    SourceType,
    TrackerEntity,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity import EntityDescription
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .config_flow import (
    CONF_SITE,
    OPT_DEVICE_TRACKER,
    OPT_SCANNED_CLIENTS,
    OPT_TRACKED_CLIENTS,
)
from .const import DOMAIN
from .controller import OmadaClientsCoordinator, OmadaSiteController

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up device trackers and scanners."""

    if not config_entry.options.get(OPT_DEVICE_TRACKER, False):
        return

    controller: OmadaSiteController = hass.data[DOMAIN][config_entry.entry_id]

    clients_coordinator = controller.get_clients_coordinator()
    site_id = config_entry.data[CONF_SITE]

    entities: list[OmadaBaseTrackerEntity] = []

    for client_mac in set().union(
        config_entry.options[OPT_SCANNED_CLIENTS],
        config_entry.options[OPT_TRACKED_CLIENTS],
    ):
        client_name = None
        try:
            # Get a display name for the client, even if the client is not currently connected
            client_name = (await controller.omada_client.get_client(client_mac)).name
        except RequestFailed as ex:
            _LOGGER.error(
                "Error getting Omada wi-fi client details for MAC %s: %s",
                client_mac,
                ex,
            )

        if client_mac in config_entry.options[OPT_SCANNED_CLIENTS]:
            entities.extend(
                [
                    OmadaClientScannerEntity(
                        site_id,
                        client_mac,
                        client_name or client_mac,
                        clients_coordinator,
                        ed,
                    )
                    for ed in DEVICE_TRACKERS
                ]
            )
        if client_mac in config_entry.options[OPT_TRACKED_CLIENTS]:
            entities.extend(
                [
                    OmadaClientTrackerEntity(
                        site_id,
                        client_mac,
                        client_name or client_mac,
                        clients_coordinator,
                        ed,
                    )
                    for ed in DEVICE_TRACKERS
                ]
            )

    async_add_entities(entities)


@dataclass(frozen=True, kw_only=True)
class OmadaDeviceTrackerEntityDescription(EntityDescription):
    """Entity description for a device tracker derived from a connected wireless client."""

    coordinator_update_func: Callable[
        [OmadaClientsCoordinator, str], OmadaWirelessClient | None
    ]
    ip_address_fn: Callable[[OmadaWirelessClient | None], str | None]
    mac_address_fn: Callable[[OmadaWirelessClient | None], str | None]
    hostname_fn: Callable[[OmadaWirelessClient | None], str | None]
    location_fn: Callable[[OmadaWirelessClient | None], str | None]
    is_connected_fn: Callable[[OmadaWirelessClient | None], bool]
    has_entity_name = True


DEVICE_TRACKERS = [
    OmadaDeviceTrackerEntityDescription(
        key="device_tracker",
        coordinator_update_func=lambda coordinator, mac: coordinator.data.get(mac),
        ip_address_fn=lambda client: client.ip if client else None,
        mac_address_fn=lambda client: client.mac if client else None,
        hostname_fn=lambda client: client.host_name if client else None,
        location_fn=lambda client: client.ap_name if client else None,
        is_connected_fn=lambda client: client.is_active if client else False,
    )
]


class OmadaBaseTrackerEntity(CoordinatorEntity[OmadaClientsCoordinator]):
    """Base entity for tracking a client connected to the Omada network."""

    entity_description: OmadaDeviceTrackerEntityDescription
    _client_details: OmadaWirelessClient | None = None

    def __init__(
        self,
        site_id: str,
        client_id: str,
        display_name: str,
        coordinator: OmadaClientsCoordinator,
        entity_description: OmadaDeviceTrackerEntityDescription,
    ) -> None:
        """Initialize the scanner."""
        super().__init__(coordinator)
        self.entity_description = entity_description
        self._site_id = site_id
        self._client_id = client_id
        self._attr_name = display_name

    @property
    def source_type(self) -> SourceType:
        """Return the source type of the device."""
        return SourceType.ROUTER

    def _do_update(self) -> None:
        self._client_details = self.entity_description.coordinator_update_func(
            self.coordinator, self._client_id
        )

    async def async_added_to_hass(self) -> None:
        """When entity is added to hass."""
        await super().async_added_to_hass()
        self._do_update()

    @callback
    def _handle_coordinator_update(self) -> None:
        """Handle updated data from the coordinator."""
        self._do_update()
        self.async_write_ha_state()


class OmadaClientScannerEntity(OmadaBaseTrackerEntity, ScannerEntity):
    """Entity for a client connected to the Omada network."""

    @property
    def ip_address(self) -> str | None:
        """Return the primary ip address of the device."""
        return self.entity_description.ip_address_fn(self._client_details)

    @property
    def mac_address(self) -> str | None:
        """Return the mac address of the device."""
        return self.entity_description.mac_address_fn(self._client_details)

    @property
    def hostname(self) -> str | None:
        """Return hostname of the device."""
        return self.entity_description.hostname_fn(self._client_details)

    @property
    def is_connected(self) -> bool:
        """Return true if the device is connected to the network."""
        return self.entity_description.is_connected_fn(self._client_details)

    @property
    def unique_id(self) -> str | None:
        """Return the unique id of the device."""
        return f"scanner_{self._site_id}_{self._client_id}"


class OmadaClientTrackerEntity(OmadaBaseTrackerEntity, TrackerEntity):
    """Entity for a client connected to the Omada network."""

    @property
    def location_name(self) -> str | None:
        """Return a location name for the current location of the device."""
        return self.entity_description.location_fn(self._client_details)

    @property
    def unique_id(self) -> str | None:
        """Return the unique id of the device."""
        return f"tracker_{self._site_id}_{self._client_details}"
