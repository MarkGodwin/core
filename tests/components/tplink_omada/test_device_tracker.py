"""Tests for TP-Link Omada device tracker entities."""

from datetime import timedelta
from unittest.mock import MagicMock

import pytest
from syrupy.assertion import SnapshotAssertion
from tplink_omada_client.exceptions import RequestFailed

from homeassistant.components.tplink_omada.config_flow import (
    OPT_DEVICE_TRACKER,
    OPT_SCANNED_CLIENTS,
    OPT_TRACKED_CLIENTS,
)
from homeassistant.components.tplink_omada.const import DOMAIN
from homeassistant.components.tplink_omada.coordinator import POLL_CLIENTS
from homeassistant.core import HomeAssistant
from homeassistant.util.dt import utcnow

from tests.common import MockConfigEntry, async_fire_time_changed

UPDATE_INTERVAL = timedelta(seconds=10)
POLL_INTERVAL = timedelta(seconds=POLL_CLIENTS + 10)

MOCK_ENTRY_DATA = {
    "host": "https://fake.omada.host",
    "verify_ssl": True,
    "site": "SiteId",
    "username": "test-username",
    "password": "test-password",
}


@pytest.fixture
async def init_integration(
    hass: HomeAssistant,
    mock_omada_clients_only_client: MagicMock,
) -> MockConfigEntry:
    """Set up the TP-Link Omada integration for testing."""
    mock_config_entry = MockConfigEntry(
        title="Test Omada Controller",
        domain=DOMAIN,
        data=dict(MOCK_ENTRY_DATA),
        options={
            OPT_DEVICE_TRACKER: True,
            OPT_TRACKED_CLIENTS: ["2E-DC-E1-C4-37-D3"],
            OPT_SCANNED_CLIENTS: ["2C-71-FF-ED-34-83"],
        },
        unique_id="12345",
    )
    mock_config_entry.add_to_hass(hass)

    await hass.config_entries.async_setup(mock_config_entry.entry_id)
    await hass.async_block_till_done()

    return mock_config_entry


async def test_device_tracker_created(
    hass: HomeAssistant,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test gateway connected switches."""

    entity_id = "device_tracker.apple"
    entity = hass.states.get(entity_id)
    assert entity is not None
    assert entity == snapshot


async def test_device_scanner_created(
    hass: HomeAssistant,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test gateway connected switches."""

    entity_id = "device_tracker.banana"
    entity = hass.states.get(entity_id)
    assert entity is not None
    assert entity == snapshot


async def test_device_scanner_created_with_default_name(
    hass: HomeAssistant,
    mock_omada_clients_only_client: MagicMock,
    mock_omada_clients_only_site_client: MagicMock,
    snapshot: SnapshotAssertion,
) -> None:
    """Test gateway connected switches."""

    mock_omada_clients_only_site_client.get_client.side_effect = RequestFailed(
        error_code=1001, msg="Some poor reason"
    )

    mock_config_entry = MockConfigEntry(
        title="Test Omada Controller",
        domain=DOMAIN,
        data=dict(MOCK_ENTRY_DATA),
        options={
            OPT_DEVICE_TRACKER: True,
            OPT_TRACKED_CLIENTS: [],
            OPT_SCANNED_CLIENTS: ["2C-71-FF-ED-34-83"],
        },
        unique_id="12345",
    )
    mock_config_entry.add_to_hass(hass)

    await hass.config_entries.async_setup(mock_config_entry.entry_id)
    await hass.async_block_till_done()

    entity_id = "device_tracker.2C_71_FF_ED_34_83"
    entity = hass.states.get(entity_id)
    assert entity is not None
    assert entity == snapshot


async def test_device_tracker_update_to_away_nulls_properties(
    hass: HomeAssistant,
    mock_omada_clients_only_site_client: MagicMock,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test gateway connected switches."""

    entity_id = "device_tracker.apple"
    entity = hass.states.get(entity_id)
    await _setup_client_disconnect(
        mock_omada_clients_only_site_client, "2E-DC-E1-C4-37-D3"
    )

    async_fire_time_changed(hass, utcnow() + POLL_INTERVAL)
    await hass.async_block_till_done()

    mock_omada_clients_only_site_client.get_connected_clients.assert_called_once()

    entity = hass.states.get(entity_id)
    assert entity is not None
    assert entity == snapshot


async def test_device_scanner_update_to_away_nulls_properties(
    hass: HomeAssistant,
    mock_omada_clients_only_site_client: MagicMock,
    init_integration: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test gateway connected switches."""

    entity_id = "device_tracker.banana"
    entity = hass.states.get(entity_id)
    await _setup_client_disconnect(
        mock_omada_clients_only_site_client, "2C-71-FF-ED-34-83"
    )

    async_fire_time_changed(hass, utcnow() + POLL_INTERVAL)
    await hass.async_block_till_done()

    mock_omada_clients_only_site_client.get_connected_clients.assert_called_once()

    entity = hass.states.get(entity_id)
    assert entity is not None
    assert entity == snapshot


async def _setup_client_disconnect(
    mock_omada_site_client: MagicMock,
    client_mac: str,
):
    filtered_clients = [
        c
        async for c in mock_omada_site_client.get_connected_clients()
        if c.mac != client_mac
    ]
    mock_omada_site_client.get_connected_clients.reset_mock()
    mock_omada_site_client.get_connected_clients.return_value = filtered_clients
