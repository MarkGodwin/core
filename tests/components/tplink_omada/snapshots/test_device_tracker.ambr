# serializer version: 1
# name: test_device_scanner_created
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Banana',
      'host_name': 'testhost',
      'ip': '192.168.1.102',
      'mac': '2C-71-FF-ED-34-83',
      'source_type': <SourceType.ROUTER: 'router'>,
    }),
    'context': <ANY>,
    'entity_id': 'device_tracker.banana',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'home',
  })
# ---
# name: test_device_scanner_created_with_default_name
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': '2C-71-FF-ED-34-83',
      'host_name': 'testhost',
      'ip': '192.168.1.102',
      'mac': '2C-71-FF-ED-34-83',
      'source_type': <SourceType.ROUTER: 'router'>,
    }),
    'context': <ANY>,
    'entity_id': 'device_tracker.2c_71_ff_ed_34_83',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'home',
  })
# ---
# name: test_device_scanner_update_to_away_nulls_properties
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Banana',
    }),
    'context': <ANY>,
    'entity_id': 'device_tracker.banana',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
# name: test_device_tracker_created
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Apple',
      'source_type': <SourceType.ROUTER: 'router'>,
    }),
    'context': <ANY>,
    'entity_id': 'device_tracker.apple',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'Spare Room',
  })
# ---
# name: test_device_tracker_update_to_away_nulls_properties
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Apple',
    }),
    'context': <ANY>,
    'entity_id': 'device_tracker.apple',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'unavailable',
  })
# ---
